### DISCUSSION TITLE
Lesson 4: Problem with deploying with Ganache CLI (Non-hexadecimal digit found)

### DISCUSSION THREAD

**Discussion by Verde888** [smartcontractkit/full-blockchain-solidity-course-py#343] [Q&A] [answered]
Hi everyone,

I had everything working well and was able to deploy transactions and see them on the Ganache GUI.
Then I installed Node.js (v16.13.0) , yarn (v1.22.17) and Ganache CLI (v6.12.2, ganache-core: 2.13.2).
Ganache CLI works OK.
I updated the deploy.py with the new address and .env with the private key, both provided by Ganache CLI.
My .env file looks like this:
`export PRIVATE_KEY=0x55ca65d1e15cd2d27678d6676e597d2690a6217a4661ae8f7a2dd9d0aa195d22`

Now, when I type "python deploy.py" in a new terminal window I get the message below.
Hard-coding my private key in the deploy.py file does not change anything.
Any idea what is happening? Thank you for your help!

 `INFO: Could not find files for the given pattern(s).
Installing solc...
Traceback (most recent call last):
  File "C:\04_PROGRAMMING\BLOCKCHAIN\demos\web3_py_simple_storage\deploy.py", line 65, in <module>
    signed_txn = w3.eth.account.sign_transaction(transaction, private_key=private_key)     
  File "C:\Users\Verde\AppData\Local\Programs\Python\Python39\lib\site-packages\eth_utils\decorators.py", line 18, in _wrapper
    return self.method(obj, *args, **kwargs)
  File "C:\Users\Verde\AppData\Local\Programs\Python\Python39\lib\site-packages\eth_account\account.py", line 727, in sign_transaction
    account = self.from_key(private_key)
  File "C:\Users\Verde\AppData\Local\Programs\Python\Python39\lib\site-packages\eth_utils\decorators.py", line 18, in _wrapper
    return self.method(obj, *args, **kwargs)
  File "C:\Users\Verde\AppData\Local\Programs\Python\Python39\lib\site-packages\eth_account\account.py", line 250, in from_key
    key = self._parsePrivateKey(private_key)
  File "C:\Users\Verde\AppData\Local\Programs\Python\Python39\lib\site-packages\eth_utils\decorators.py", line 18, in _wrapper
    return self.method(obj, *args, **kwargs)
  File "C:\Users\Verde\AppData\Local\Programs\Python\Python39\lib\site-packages\eth_account\account.py", line 774, in _parsePrivateKey
    return self._keys.PrivateKey(HexBytes(key))
  File "C:\Users\Verde\AppData\Local\Programs\Python\Python39\lib\site-packages\hexbytes\main.py", line 23, in __new__
    bytesval = to_bytes(val)
  File "C:\Users\Verde\AppData\Local\Programs\Python\Python39\lib\site-packages\hexbytes\_utils.py", line 17, in to_bytes
    return hexstr_to_bytes(val)
  File "C:\Users\Verde\AppData\Local\Programs\Python\Python39\lib\site-packages\hexbytes\_utils.py", line 50, in hexstr_to_bytes
    return binascii.unhexlify(ascii_hex)
binascii.Error: Non-hexadecimal digit found`

**Comment by cromewar** [comment]
there might be a problem with the bytcode encode on your deploy script, please copy and paste it here, I'll check it.

**Comment by Verde888** [comment]
Hi, thank you in advance. 
Do you mean the deploy.py file? Here it is:

`
```Python
from solcx import compile_standard, install_solc
import json
from web3 import Web3
import os
from dotenv import load_dotenv

load_dotenv()

with open("./SimpleStorage.sol", "r") as file:
    simple_storage_file = file.read()


print("Installing solc...")
install_solc("0.6.0")


compiled_sol = compile_standard(
    {
        "language": "Solidity",
        "sources": {"SimpleStorage.sol": {"content": simple_storage_file}},
        "settings": {
            "outputSelection": {
                "*": {"*": ["abi", "metadata", "evm.bytecode", "evm.sourceMap"]}
            }
        },
    },
    solc_version="0.6.0",
)

with open("compiled_code.json", "w") as file:
    json.dump(compiled_sol, file)


bytecode = compiled_sol["contracts"]["SimpleStorage.sol"]["SimpleStorage"]["evm"][
    "bytecode"
]["object"]


abi = compiled_sol["contracts"]["SimpleStorage.sol"]["SimpleStorage"]["abi"]


w3 = Web3(Web3.HTTPProvider("http://127.0.0.1:8545"))
chain_id = 1337
my_address = "0x3bDf1eEcce4162f0C751C322dc3BFb4DBB8cd646"
private_key = os.getenv("PRIVATE_KEY")


SimpleStorage = w3.eth.contract(abi=abi, bytecode=bytecode)


nonce = w3.eth.getTransactionCount(my_address)


transaction = SimpleStorage.constructor().buildTransaction(
    {"chainId": chain_id, "from": my_address, "nonce": nonce}
)

signed_txn = w3.eth.account.sign_transaction(transaction, private_key=private_key)


print("Deploying contract...")
tx_hash = w3.eth.send_raw_transaction(signed_txn.rawTransaction)
tx_receipt = w3.eth.wait_for_transaction_receipt(tx_hash)  ## Wait for confirmation
print("Contract deployed!")



simple_storage = w3.eth.contract(address=tx_receipt.contractAddress, abi=abi)


print("Initial value stored on the contract:")
print(simple_storage.functions.retrieve().call())
print("Updating contract...")


store_transaction = simple_storage.functions.store(15).buildTransaction(
    {"chainId": chain_id, "from": my_address, "nonce": nonce + 1}
)

signed_store_txn = w3.eth.account.sign_transaction(
    store_transaction, private_key=private_key
)

send_store_txn = w3.eth.send_raw_transaction(signed_store_txn.rawTransaction)
tx_receipt = w3.eth.wait_for_transaction_receipt(send_store_txn)
print("Contract updated! New value stored on the contract:")
print(simple_storage.functions.retrieve().call())
`
```

**Comment by cromewar** [comment, marked as answer]
Your code seems to be ok, the problem most probably is on `dotenv`python package which is nor recognizing your private key. Try to follow this steps:

1. use a python virtualenv and create a new instance.
2. install [this](https://pypi.org/project/python-dotenv/) package.
3. add your private key to Windows environment variables.
4. try again.

**Reply by Verde888** [reply]
Hi, thank you. I installed everything in the venv this time. It works fine if I hardcode the private key in the script - it deploys the contract.

But retrieving the key from the .env does not work, still returns the same error (adding the environment variable in Windows System Properties did not change anything).
At least I have the possibility to continue with the course with a hardcoded key... meanwhile I will probably move to WSL...

If anyone knows where the bug can come from, please help!

**Reply by cromewar** [reply]
You should reload powershell or cmd after adding the new environment variable or it won't be recognized... and yes using WSL is better idea, you can export env variables just using `export`command.

I'll set this thread as solved, please post any other troubles or question you have of the course, best regards.

**Reply by shucks18** [reply]
Hey @Verde888 did you manage to solve the problem cause I'm getting the same issue, and also how is WSL?

**Reply by cromewar** [reply]
@Verde888 is solved? and @shucks18 you can see info for wsl [Here](https://medium.com/@cromewar/how-to-setup-windows-10-11-for-smart-contract-development-and-brownie-e7d8d13555b3).

**Reply by shucks18** [reply]
cheers man appreciate it! @cromewar
